src/generic/eigen_solver.cc

```diff
@@ -700,19 +736,205 @@ namespace oomph
+
+    // Brute force computation of eigenvalues (allowing for NaNs and Infs)
+    unsigned n = problem_pt->ndof();
+    eigenvalue.resize(n);
+    for (unsigned i = 0; i < n; ++i)
+    {
+      // N.B. This is naive and dangerous according to the documentation
+      // beta could be very small giving over or under flow
+      // Shift has already been taken out in helper function
+      eigenvalue[i] = std::complex<double>(alpha_eval[i].real() / beta_eval[i],
+                                           alpha_eval[i].imag() / beta_eval[i]);
+    }
+  }
+
+  //==========================================================================
+  /// Solve the real eigenproblem that is assembled by elements in
+  /// a mesh in a Problem object. Note that the assembled matrices include the
+  /// shift and are real. The eigenvalues and eigenvectors are,
+  /// in general, complex. Eigenvalues may be infinite and are therefore
+  /// returned as
+  /// \f$ \lambda_i = \alpha_i / \beta_i \f$ where \f$ \alpha_i \f$ is complex
+  /// while \f$ \beta_i \f$ is real. The actual eigenvalues may then be
+  /// computed by doing the division, checking for zero betas to avoid NaNs.
+  /// There's a convenience wrapper to this function that simply computes
+  /// these eigenvalues regardless. That version may die in NaN checking is
+  /// enabled (via the fenv.h header and the associated feenable function).
+  /// At least n_eval eigenvalues are computed.
+  //==========================================================================
+  void LAPACK_QZ::solve_eigenproblem(
+    Problem* const& problem_pt,
+    const int& n_eval,
+    Vector<std::complex<double>>& alpha_eval,
+    Vector<double>& beta_eval,
+    Vector<Vector<std::complex<double>>>& eigenvector)
+  {
+    Vector<DoubleVector> eigenvector_aux;
+
+    // Call raw interface to lapack qz
+    solve_eigenproblem_helper(
+      problem_pt, n_eval, alpha_eval, beta_eval, eigenvector_aux);
+
+    // Now move auxiliary data into actual complex eigenvalues
+    // Instrutions from lapack qz (where VR translates into eigenvector_aux):
+    //        VR is DOUBLE PRECISION array, dimension (LDVR,N)
+    //        If JOBVR = 'V', the right eigenvectors v(j) are stored one
+    //        after another in the columns of VR, in the same order as
+    //        their eigenvalues. If the j-th eigenvalue is real, then
+    //        v(j) = VR(:,j), the j-th column of VR. If the j-th and
+    //        (j+1)-th eigenvalues form a complex conjugate pair, then
+    //        v(j) = VR(:,j)+i*VR(:,j+1) and v(j+1) = VR(:,j)-i*VR(:,j+1).
+    //        Each eigenvector is scaled so the largest component has
+    //        abs(real part)+abs(imag. part)=1.
+    unsigned n = problem_pt->ndof();
+    eigenvector.resize(n);
+    unsigned eval_count = 0;
+    while (eval_count < n)
+    {
+      // i-th eigenvalue is real:
+      if (alpha_eval[eval_count].imag() == 0.0)
```

Comment:
Should there be a threshold in here given that this is a floating point comparison?

Reply:
Don't think so, According the lapack's own documentation (copied into the code for clarity) the distinction is made based on complexness, not near-complexness. 

 Done.